go/service/gregor_test.go

```diff
@@ -173,3 +182,206 @@ func TestShowTrackerPopupMessage(t *testing.T) {
+	if state, err = h.gregorCli.StateMachineState(nil); err != nil {
+		t.Fatal(err)
+	}
+
+	items, err := state.Items()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(items) != size {
+		t.Fatalf(message, len(items), size)
+	}
+}
+
+func TestSyncBasic(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor")
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 20
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		server.ConsumeMessage(context.TODO(), msg)
+	}
+
+	// Sync messages down and see if we get 20
+	var syncedMsgs int
+	var err error
+	if syncedMsgs, err = h.reSync(context.TODO(), server); err != nil {
+		t.Fatal(err)
+	}
+
+	if syncedMsgs != numMsgs {
+		t.Fatalf("synced messages does not equal consumed, %d != %d", syncedMsgs, numMsgs)
+	}
+
+	checkStateSize(t, "state items not equal to consumed, %d != %d", h, numMsgs)
+}
+
+func TestSyncBroadcast(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor")
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 6
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		server.ConsumeMessage(context.TODO(), msg)
+		if i < numMsgs/2 {
+			h.BroadcastMessage(context.TODO(), msg)
+		}
+	}
+
+	// Turn off fresh sync
+	h.freshSync = false
+
+	// We should only get half of the messages on a non-fresh sync
+	var syncedMsgs int
+	var err error
+	if syncedMsgs, err = h.reSync(context.TODO(), server); err != nil {
+		t.Fatal(err)
+	}
+
+	if syncedMsgs != numMsgs/2+1 {
+		t.Fatalf("syncMsgs should be half of numMsgs, %d, %d", syncedMsgs, numMsgs/2+1)
+	}
+
+	checkStateSize(t, "state items not equal to consumed, %d != %d", h, numMsgs)
+}
+
+func TestSyncSaveRestore(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor")
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 6
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		server.ConsumeMessage(context.TODO(), msg)
+		if i < numMsgs/2 {
+			h.BroadcastMessage(context.TODO(), msg)
+		}
+	}
+
+	// Try saving
+	var err error
+	if err = h.gregorCli.Save(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a new gregor handler, this will restore our saved state
+	if h, err = newGregorHandler(tc.G); err != nil {
+		t.Fatal(err)
+	}
+
+	checkStateSize(t, "restore brought back the wrong number of items: %d != %d", h, numMsgs/2)
+
+	// Sync from the server
+	var syncedMsgs int
+	if syncedMsgs, err = h.reSync(context.TODO(), server); err != nil {
+		t.Fatal(err)
+	}
+	if syncedMsgs != numMsgs {
```

Comment:
I think if that is wrong then we will fail the hash check, which would make our numbers wrong.